api/outdated/outdated/tests/test_validators.py

```diff
@@ -0,0 +1,170 @@
+        assert access_token
+
+
+@pytest.mark.parametrize("access_token", [None, "token"])
+@pytest.mark.parametrize("is_public", [True, False])
+@pytest.mark.parametrize("repo_type", ["public", "access-token"])
+@pytest.mark.parametrize("remote_exists", [True, False])
+def test_remote_exists(  # noqa: C901
+    repo_type: str,
+    access_token: str | None,
+    is_public: bool,
+    remote_exists: bool,
+    mocker: MockerFixture,
+) -> None:
+    def side_effect(url: str) -> bool:
+        if not remote_exists:
+            return False
+        if repo_type == "public":
+            return True
+        return is_public or "@" in url
+
+    check_remote_existance_mock = mocker.patch.object(
+        validators, "check_remote_existance", side_effect=side_effect
+    )
+
+    attrs = {
+        "repo": "my.git.com/foo/bar",
+        "repo_type": repo_type,
+    }
+
+    if access_token:
+        attrs["access_token"] = access_token
+
+    try:
+        validators.validate_remote_url(attrs, ProjectSerializer())
+        if repo_type == "access-token" and access_token:
+            check_remote_existance_mock.assert_has_calls(
+                [call("outdated:token@my.git.com/foo/bar"), call("my.git.com/foo/bar")]
+            )
+            assert remote_exists
+            return
+        if (
+            repo_type == "access-token"
+            and not access_token
+            or repo_type == "public"
+            and access_token
+        ):
+            check_remote_existance_mock.assert_not_called()
+            return
+        check_remote_existance_mock.assert_called_once()
+        assert remote_exists
+    except ValidationError as e:
+        if not remote_exists:
+            assert e.message == "Repository does not exist."  # noqa: PT017
+            return
+        assert e.message == "Repository is public."  # noqa: PT017
+        assert is_public
+        assert access_token
+        assert repo_type == "access-token"
+
+
+@pytest.mark.django_db()
+@pytest.mark.parametrize("repo", ["my.git.com/foo/bar", "other.git.com/foo/bar"])
+@pytest.mark.parametrize("repo_type", ["public", "access-token"])
+def test_remote_exists_with_instance(
+    repo_type: RepoType,
+    repo: str,
+    mocker: MockerFixture,
+    project_factory: ProjectFactory,
+) -> None:
+    check_remote_existance_mock = mocker.patch.object(
+        validators, "check_remote_existance"
+    )
+
+    project = project_factory(repo="my.git.com/foo/bar", repo_type="public")
+
+    attrs = {"repo": repo, "repo_type": repo_type}
+
+    if repo_type == "access-token":
+        attrs["access_token"] = "token"  # noqa: S105
+
+    serializer = ProjectSerializer(project)
+
+    with suppress(ValidationError):
+        validators.validate_remote_url(attrs, serializer)
+
+    if project.repo == repo and project.repo_type == repo_type:
+        check_remote_existance_mock.assert_not_called()
+    else:
+        check_remote_existance_mock.assert_called()
```

Comment:
Not sure if I understand this 100% right.

```python
    try:
        validators.validate_remote_url(attrs, serializer)
    except ValidationError:
        assert project.repo == repo and project.repo_type == repo_type
        check_remote_existance_mock.assert_not_called()
    else:
        assert project.repo != repo and project.repo_type != repo_type
        check_remote_existance_mock.assert_called()
```

Reply:
i have rewritten this test